File: runtime/Cpp/runtime/src/LexerInterpreter.cpp
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

#include "atn/ATNType.h"
#include "atn/LexerATNSimulator.h"
#include "dfa/DFA.h"
#include "Exceptions.h"
#include "Vocabulary.h"

#include "LexerInterpreter.h"

using namespace antlr4;

LexerInterpreter::LexerInterpreter(std::string_view grammarFileName, const dfa::Vocabulary &vocabulary,
  antlrcpp::Span<const std::string_view> ruleNames, antlrcpp::Span<const std::string_view> channelNames, antlrcpp::Span<const std::string_view> modeNames,
  const atn::ATN &atn, CharStream *input)
  : Lexer(input), _grammarFileName(grammarFileName), _atn(atn), _ruleNames(ruleNames),
                  _channelNames(channelNames), _modeNames(modeNames),
                  _vocabulary(vocabulary) {

  if (_atn.grammarType != atn::ATNType::LEXER) {
    throw IllegalArgumentException("The ATN must be a lexer ATN.");
  }

  for (size_t i = 0; i < atn.getNumberOfDecisions(); ++i) {
    _decisionToDFA.push_back(dfa::DFA(_atn.getDecisionState(i), i));
  }
  _interpreter = new atn::LexerATNSimulator(this, _atn, antlrcpp::Span<antlr4::dfa::DFA>(_decisionToDFA), _sharedContextCache); /* mem-check: deleted in d-tor */
}

LexerInterpreter::~LexerInterpreter()
{
  delete _interpreter;
}

const atn::ATN& LexerInterpreter::getATN() const {
  return _atn;
}

std::string_view LexerInterpreter::getGrammarFileName() const {
  return _grammarFileName;
}

antlrcpp::Span<const std::string_view> LexerInterpreter::getRuleNames() const {
  return _ruleNames;
}

antlrcpp::Span<const std::string_view> LexerInterpreter::getChannelNames() const {
  return _channelNames;
}

antlrcpp::Span<const std::string_view> LexerInterpreter::getModeNames() const {
  return _modeNames;
}

const Vocabulary& LexerInterpreter::getVocabulary() const {
  return _vocabulary;
}

File: runtime/Swift/Sources/Antlr4/LexerInterpreter.swift
/// 
/// Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
/// Use of this file is governed by the BSD 3-clause license that
/// can be found in the LICENSE.txt file in the project root.
/// 


public class LexerInterpreter: Lexer {
    internal let grammarFileName: String
    internal let atn: ATN

    internal let ruleNames: [String]
    internal let channelNames: [String]
    internal let modeNames: [String]

    private let vocabulary: Vocabulary?

    internal final var _decisionToDFA: [DFA]
    internal let _sharedContextCache = PredictionContextCache()

    public init(_ grammarFileName: String, _ vocabulary: Vocabulary, _ ruleNames: Array<String>, _ channelNames: Array<String>, _ modeNames: Array<String>, _ atn: ATN, _ input: CharStream) throws {

        self.grammarFileName = grammarFileName
        self.atn = atn
        self.ruleNames = ruleNames
        self.channelNames = channelNames
        self.modeNames = modeNames
        self.vocabulary = vocabulary

        self._decisionToDFA = [DFA]()
        for i in 0 ..< atn.getNumberOfDecisions() {
            _decisionToDFA[i] = DFA(atn.getDecisionState(i)!, i)
        }
        super.init(input)
        self._interp = LexerATNSimulator(self, atn, _decisionToDFA, _sharedContextCache)

        if atn.grammarType != ATNType.lexer {
            throw ANTLRError.illegalArgument(msg: "The ATN must be a lexer ATN.")

        }
    }

    public required init(_ input: CharStream) {
        fatalError("Use the other initializer")
    }

    override
    public func getATN() -> ATN {
        return atn
    }

    override
    public func getGrammarFileName() -> String {
        return grammarFileName
    }

    override
    public func getRuleNames() -> [String] {
        return ruleNames
    }

    override
    public func getChannelNames() -> [String] {
        return channelNames
    }

    override
    public func getModeNames() -> [String] {
        return modeNames
    }

    override
    public func getVocabulary() -> Vocabulary {
        if vocabulary != nil {
            return vocabulary!
        }

        return super.getVocabulary()
    }
}
